### PR TITLE
Feature/multithreading revamp

### PR DESCRIPTION
**Pull request by neuneck**
I revamped the threaded evaluation and employ the ThreadPoolExecutor from python's concurrent.futures module. This module was added in python3, but a backport for python2 is available on PYPI as "futures" package.

The ThreadPoolExecutor has a few key advantages over manual Thread management:
* It makes sure all threads are cleaned up, even if a exception breaks the loop
* It allows to set a maximum number of threads (useful for very wide graphs)
* It raises exceptions that happen in threads spawned by it with a useful stacktrace.

I have not yet added the futures package to the requirements.txt, since it might break python3 installations. The recommended method (cf. https://pypi.org/project/futures/) to add it to a project is by adding it to the setup.py installation requirements for python2. 

This PR closes https://github.com/PaulSchweizer/flowpipe/issues/112 as the situation in question can no longer occur and exceptions in threads are handled pythonically.

### CODE REVIEW

File: flowpipe/graph.py [hunk 1]
```diff
@@ -205,7 +205,7 @@ def accepts_connection(self, output_plug, input_plug):
         return True
 
     def evaluate(self, mode="linear", skip_clean=False,
-                 submission_delay=0.1, raise_after=None):
+                 submission_delay=0.1, max_workers=None):
```
Comment (PaulSchweizer) [inline flowpipe/graph.py]: Nice, the max workers was definitely missing, thanks for introducing it

File: flowpipe/graph.py [hunk 2]
```diff
@@ -228,19 +228,21 @@ def evaluate(self, mode="linear", skip_clean=False,
...
+            "threading": (self._evaluate_threaded, {"max_workers": max_workers}),
+            "multiprocessing": (self._evaluate_multiprocessed,
+                                {"submission_delay": submission_delay})
         }
 
         try:
-            eval_func = eval_modes[mode]
+            eval_func, eval_func_args = eval_modes[mode]
         except KeyError:
             mode_options = ""
```
Comment (PaulSchweizer) [inline flowpipe/graph.py]: Not really part of the PR but I saw this recently, this could be solved more elegantly with join, like this:
```python
raise ValueError("Invalid mode {0}, options are {1}".format(
                mode, ", ".join(list(eval_modes.keys()))))
```

File: flowpipe/graph.py [hunk 3]
```diff
@@ -249,88 +251,68 @@ def evaluate(self, mode="linear", skip_clean=False,
...
+        """Evaluate each node in a new thread."""
+        log.debug("{0} evaluating {1} nodes in threading mode.".format(
+            self.name, len(nodes_to_evaluate)))
+        def node_runner(node):
+            """Run a node's evaluate method and return the node."""
+            node.evaluate()
+            return node
+
+        running_futures = []
+        with futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
```
Comment (PaulSchweizer) [inline flowpipe/graph.py]: Cool, I did not know about this

File: flowpipe/graph.py [hunk 4]
```diff
@@ -5,10 +5,10 @@
     from collections import OrderedDict
 except ImportError:
     from ordereddict import OrderedDict
+from concurrent import futures
```
Comment (PaulSchweizer) [inline flowpipe/graph.py]: This still needs to be added to the requirements

Comment (PaulSchweizer) [reply flowpipe/graph.py]: Sorry, ignore that, had overlooked your comment

Comment (PaulSchweizer) [reply flowpipe/graph.py]: Let's add it to the setup.py as described in your link and see if the tests on travis run through

**Comment by neuneck** [comment]
@PaulSchweizer I just realized I merged this before your approval. Sorry about that, I was merging PRs at work constantly and lost track of the approval state of this one apparently.

If you have further comments/improvement suggestions, please do add them and I will take care of it ASAP.

**Comment by PaulSchweizer** [comment]
@neuneck , no worries, I thought it was an accident anyways, stuff like that just happens from time to time :D Thanks for contributing!
